pennylane/templates/subroutines/fermionic_double_excitation.py

```diff
@@ -32,7 +32,7 @@
 four_term_grad_recipe = ([[c1, 1, a], [-c1, 1, -a], [-c2, 1, b], [c2, 1, -b]],)
 
 
-def _layer1(weight, s, r, q, p, set_cnot_wires):
+def _layer1(weight, s, r, q, p, set_cnot_wires, op_list):
```

Comment:
I think this can be avoided; we can access `_layer1` and `_layer2` etc. using the operation recorder:

```pycon
>>> def layer1():
...     qml.Hadamard(wires=0)
...     qml.RZ(0.5, wires=0)
...     qml.AngleEmbedding([0.1, 0.2, 0.3], wires=[0, 1, 2])
>>> with qml.tape.OperationRecorder() as q:
...     layer1()
>>> q.operations
[Hadamard(wires=[0]),
 RZ(0.5, wires=[0]),
 AngleEmbedding([0.1, 0.2, 0.3], wires=[0, 1, 2])]
```

This is also safe, unlike using a tape, as operation recorders do not append themselves to queues.

Reply:
In fact I think this is a better approach than passing an argument to a function to be mutated, which is a side effect, and can (a) break the JIT, and (b) be a cause of bugs

Reply:
This is a strong argument. But it makes operator design depend on an obscure functionality that we wanted to deprecate several times if I remember correctly? It also ties the idea of tapes into the representations, which this PR is exactly there to avoid.

Let's do something else, the subfunction creates its own list and returns is, which is merged into the `op_list` one in the decomposition. Much cleaner, should have done this already!

Reply:
I must admit, I was also under the impression that we wanted to deprecate the `OperationRecorder`, so when I set down to write the comment above I tried to write it without using the `OperationRecorder`. But... I couldn't do it, it seems that the `OperationRecorder` is quite unique in functionality?

- It tracks contained operations
- Apart from that it doesn't do anything; it doesn't interfere with surrounding tapes, it doesn't interact with the queue

And now I can't recall why we wanted to deprecate it in the first place 🤔